src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionHResultMarshaller.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace System.Runtime.InteropServices.Marshalling
+{
+    /// <summary>
+    /// Marshals an exception object to the value of its <see cref="Exception.HResult"/> converted to <typeparamref name="T"/>.
+    /// </summary>
+    /// <typeparam name="T">The unmanaged type to convert the HResult to.</typeparam>
+    [CustomMarshaller(typeof(Exception), MarshalMode.UnmanagedToManagedOut, typeof(ExceptionDefaultMarshaller<>))]
+    public static class ExceptionHResultMarshaller<T>
+        where T : unmanaged, INumber<T>
+    {
+        public static T ConvertToUnmanaged(Exception e)
+        {
+            // Use GetHRForException to ensure the runtime sets up the IErrorInfo object
+            // and calls SetErrorInfo if the platform suppots it.
+            // TODO: Should we use the built-in COM interop support for this, or should we use the generator to implement
+            // this experience?
+            return T.CreateTruncating(Marshal.GetHRForException(e));
```

Comment:
I don't understand this `CreateTruncating` call.

Reply:
It does help, but this is an odd way to express this logic. From what I am seeing here, this code is also applied to `double` and `float`, which means this is no longer an `HRESULT`. I'm also not sure consuming `INumber` here is a good idea as I think it will root all `INumber` interfaces for a COM scenario regardless if this is the only place it is used. Why is the generic needed for `HRESULT`?

Reply:
I'm using the generic math here to deduplicate a little code. The `ExceptionHResultMarshaller<T>` is meant to cover the cases where `int` and `uint` are used as the unmanaged signature return type. For `ExceptionNaNMarshaller<T>` is meant to cover `float` and `double`. I can switch from the generic math to separate types, but I was trying to avoid code duplication now that we have these interfaces available.

Reply:
I think this is the last query I have.

Reply:
I think the tradeoff is worth it. We can explore breaking it up and duplicating the APIs when we take these to API review.